### PR TITLE
Start the architecture section

### PR DESCRIPTION
**Pull request by asmeurer**
I didn't get LaTeX installed yet and it's been awhile since I've used it, so don't be surprised if this doesn't compile.

### CODE REVIEW

File: paper.tex [hunk 1]
```diff
@@ -15,6 +15,53 @@ \section{Architecture}
...
+>>> expr.args[0]
+1
+>>> type(expr.args[0])
+<class 'sympy.core.numbers.One'>
+>>> expr.args[0].args
+()
+\end{verbatim}
+
+Every SymPy expression satisfies a key invariant, namely,
+\tt{expr.func(*expr.args) == expr}. This means that expressions rebuildable
```
Comment (Upabjojr) [inline paper.tex]: Need to specify that SymPy expressions are immutable.

File: paper.tex [hunk 2]
```diff
@@ -15,6 +15,53 @@ \section{Architecture}
...
+from their \tt{args}. \footnote{\tt{expr.func} is used instead of
+  \tt{type(expr)} to allow the function of an expression to be distinct from
+  its actual Python class. In most cases the two are the same.} Here, we note
```
Comment (Upabjojr) [inline paper.tex]: example of `AppliedPredicate`?

File: paper.tex (hunk 1 above)
Comment (asmeurer) [inline paper.tex]: Yeah, I know. I was unsure where exactly to put this in there.

**Comment by asmeurer** [comment]
Anyone have any idea what this error means?

```
This is pdfTeX, Version 3.14159265-2.6-1.40.15 (TeX Live 2014/Fink) (preloaded format=pdflatex)
 restricted \write18 enabled.
entering extended mode
(./paper.tex
LaTeX2e <2014/05/01>
Babel <3.9k> and hyphenation patterns for 78 languages loaded.
(/sw/share/texmf-dist/tex/latex/base/article.cls
Document Class: article 2007/10/19 v1.4h Standard LaTeX document class
(/sw/share/texmf-dist/tex/latex/base/size10.clo)) (./paper.aux)
Overfull \hbox (27.99953pt too wide) in paragraph at lines 55--63
[]\OT1/cmr/m/n/10 Every SymPy ex-pres-sion sat-is-fies a key in-vari-ant, namel
y, \OT1/cmtt/m/n/10 expr.func(*expr.args)
! Missing $ inserted.
<inserted text>
                $
l.66 \texttt{(x.__mul__(y)).__add__(2)}
                                       . \texttt{x} and \texttt{y}, returned...
```

**Comment by Upabjojr** [comment]
Underscores aren't special operators in LaTeX?

**Comment by asmeurer** [comment]
Yeah, I guess they are. I didn't realize they mattered outside of math.

**Comment by Upabjojr** [comment]
Use `\verb` instead of `\_`.

File: paper.tex [hunk 3]
```diff
@@ -15,6 +15,77 @@ \section{Architecture}
...
+  distinct from its actual Python class. In most cases the two are the same.}.
+Here, we note that in SymPy, the \texttt{==} operator represents exact
+structural equality, not mathematical equality. This allows us to test if any
+two expressions are equal to one another as expression trees.
+
+Python allows classes to overload operators. The Python interpreter translates
+above \texttt{x*y + 2} to, roughly,
+\texttt{(x.\_\_mul\_\_(y)).\_\_add\_\_(2)}. \texttt{x} and \texttt{y}, returned from
+the \texttt{symbols} function, are \texttt{Symbol} instances. The \texttt{2}
+in the expression is processed by Python as a literal, and is stored as
```
Comment (asmeurer) [inline paper.tex]: Is there a better way to write this? We _don't_ store it as an `int`.

**Comment by Upabjojr** [comment]
http://tex.stackexchange.com/questions/2790/when-should-one-use-verb-and-when-texttt

**Comment by asmeurer** [comment]
I can't get the \verb to work in the footnote. If you can make it work you can send a PR to this branch (or just push directly).

**Comment by asmeurer** [comment]
I agree it's better, though, at least for the present formatting, `\verb` uses a better underscore than `\texttt{\_}`.

File: paper.tex (hunk 3 above)
Comment (Upabjojr) [inline paper.tex]: stored in a class _Integer_, which is a wrapper around Python's builtin type.

Comment (asmeurer) [inline paper.tex]: The point I want to get across here is that even though the `2` is interpreted by Python as an `int`, it gets converted to Integer when it gets passed in to the arguments of `__add__`.

Comment (asmeurer) [inline paper.tex]: Maybe "stored" -> "parsed"

**Comment by certik** [comment]
Let's fix #20 before merging this, to ensure things don't break for other people.

**Comment by asmeurer** [comment]
I guess \verb also silences some chktex warnings about spaces after parentheses.

**Comment by certik** [comment]
(Closed & reopened to trigger a Travis build.)

**Pull request merged**